homeassistant/components/onvif/camera.py

```diff
@@ -103,59 +125,130 @@ def __init__(self, hass, config):
         self._name = config.get(CONF_NAME)
         self._ffmpeg_arguments = config.get(CONF_EXTRA_ARGUMENTS)
         self._profile_index = config.get(CONF_PROFILE)
+        self._ptz_service = None
         self._input = None
-        self._media_service = \
-            onvif.ONVIFService('http://{}:{}/onvif/device_service'.format(
-                self._host, self._port),
-                               self._username, self._password,
-                               '{}/wsdl/media.wsdl'.format(os.path.dirname(
-                                   onvif.__file__)))
-
-        self._ptz_service = \
-            onvif.ONVIFService('http://{}:{}/onvif/device_service'.format(
-                self._host, self._port),
-                               self._username, self._password,
-                               '{}/wsdl/ptz.wsdl'.format(os.path.dirname(
-                                   onvif.__file__)))
-
-    def obtain_input_uri(self):
+
+        _LOGGER.debug("Setting up the ONVIF camera device @ '%s:%s'",
+                      self._host,
+                      self._port)
+
+        self._camera = ONVIFCamera(self._host,
+                                   self._port,
+                                   self._username,
+                                   self._password,
+                                   '{}/wsdl/'
+                                   .format(os.path.dirname(onvif.__file__)))
+
+    async def async_initialize(self):
+        """
+        Initialize the camera.
+
+        Initializes the camera by obtaining the input uri and connecting to
+        the camera. Also retrieves the ONVIF profiles.
+        """
+        _LOGGER.debug("Updating service addresses")
+
+        await self._camera.update_xaddrs()
```

Comment:
If we can't connect, let's raise `PlatformNotReady`

```
from homeassistant.exceptions import PlatformNotReady
```

Home Assistant will automatically retry setting it up in the future. Note that we should not raise this if we know authentication is invalid. We should only raise this if the camera is not currently reachable.

Reply:
From initial testing, a connection error will throw `aiohttp.client_exceptions.ClientConnectorError` and invalid authorization will throw `zeep.exceptions.Fault` with the message "The action requested requires authorization and the sender is not authorized".

Would probably be good to catch both of those and log an error.  Also, I just released 0.1.2 which contains some related fixes.

Reply:
Updated the lib, and checking for both exceptions (with logging). Will raise PlatformNotReady when the exception is ClientConnectorError.

I am new to Python, so please double check whether I did implement this correctly.